handel/src/aggregation.rs

```diff
@@ -326,100 +332,190 @@ impl<
+        // special case of full contributions
+        if todo.level == self.protocol.partitioner().levels()
+            && self.is_complete_aggregate(&todo.contribution)
+        {
+            return todo.contribution;
+        }
+
+        // if the contribution is valid push it to the store, creating a new aggregate
+        {
+            let store = self.protocol.store();
+            let mut store = store.write();
+            store.put(
+                todo.contribution.clone(),
+                todo.level,
+                self.protocol.registry(),
+                self.protocol.identify(),
+            );
+        }
+
+        // in case the level of this todo has not started, start it now as we have already contributions on it.
+        self.start_level(todo.level);
+        // check if a level was completed by the addition of the contribution
+        self.check_completed_level(todo.level);
+
+        // get the best aggregate
+        let last_level = self.levels.last().expect("No levels");
+        let best = {
+            let store = self.protocol.store();
+            let store = store.read();
+            store.combined(last_level.id)
+        };
+
+        best.expect("A best signature must exist after applying a contribution.")
+    }
+
+    /// Verifies a given signature.
+    ///
+    /// As signature verification can be costly it creates a future which will be polled directly after creation.
+    /// The created future may return immediately.
+    ///
+    /// This function will override `self.current_verification` if the created future does not immediately produce a value.
+    ///
+    /// ## Returns
+    /// Returns the verification result and the todo in question iff the verification resolves immediately.
+    /// None otherwise
+    fn start_todo_verification(
+        &mut self,
+        todo: TodoItem<P::Contribution>,
+        cx: &mut Context<'_>,
+    ) -> Option<(VerificationResult, TodoItem<P::Contribution>)> {
+        // Create a new verification future.
+        let verifier = self.protocol.verifier();
+        let mut fut = async move {
+            let result = verifier.verify(&todo.contribution).await;
+            (result, todo)
+        }
+        .boxed();
+
+        if let Poll::Ready(result) = fut.poll_unpin(cx) {
+            return Some(result);
+        }
+
+        self.current_verification = Some(fut);
+        None
+    }
+}
+
+impl<TId, P, N> Stream for OngoingAggregation<TId, P, N>
+where
+    TId: Debug + Clone + Unpin + Send + 'static,
+    P: Protocol<TId>,
+    N: Network<Contribution = P::Contribution>,
+{
+    type Item = P::Contribution;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Check if the automatic update interval triggers, if so perform the update.
+        if let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```

Comment:
```suggestion
        while let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```
This needs to be a `while let`, otherwise the interval doesn't register itself with the waker if `Poll::Ready` is returned: https://docs.rs/tokio/1.38.0/tokio/time/struct.Interval.html#method.poll_tick.